### PR TITLE
Support model extensibility (but not polymorphism via 'discriminator' yet)

### PR DESCRIPTION
**Pull request by captin411**
Adds support for polymorphic models as mentioned in Yelp/bravado-core#7

### CODE REVIEW

**Comment by analogue** [comment]
Thanks for the PR. Will find some time to review this next week.

File: bravado_core/schema.py [hunk 1]
```diff
@@ -93,3 +96,34 @@ def get_spec_for_prop(swagger_spec, object_spec, object_value, prop_name):
     raise SwaggerMappingError(
         "Don't know what to do with `additionalProperties` in spec {0} "
         "when inspecting value {1}".format(object_spec, object_value))
+
+
+def collapsed_properties(model_spec, swagger_spec):
```
Comment (analogue) [inline bravado_core/schema.py]: I'd very much like to see unit tests for this method in `collapsed_properties_test.py`.

File: bravado_core/model.py [hunk 2]
```diff
@@ -88,23 +89,28 @@ def create_model_type(swagger_spec, model_name, model_spec):
...
     plus additional attibutes that exist as `additionalProperties`.
 
     :param model: instance of a model
     :param model_spec: spec the passed in model in dict form
+    :type swagger_spec: :class:`bravado_core.spec.Spec`
     :returns: list of str
     """
-    return list(model_spec['properties'].keys()) + model._additional_props
+    return (list(collapsed_properties(model_spec, swagger_spec).keys()) +
+            model._additional_props)
```
Comment (analogue) [inline bravado_core/model.py]: Outermost parens seem extraneous

File: bravado_core/model.py [hunk 3]
```diff
@@ -137,6 +143,8 @@ def model_constructor(model, model_spec, constructor_kwargs):
     :type model: type
     :param model_spec: model specification
     :type model_spec: dict
+    :param swagger_spec: :class:`bravado_core.spec.Spec`
```
Comment (analogue) [inline bravado_core/model.py]: Remove - duplicates :type:

File: bravado_core/model.py [hunk 4]
```diff
@@ -144,7 +152,9 @@ def model_constructor(model, model_spec, constructor_kwargs):
     """
     arg_names = list(constructor_kwargs.keys())
 
-    for attr_name, attr_spec in iteritems(model_spec['properties']):
+    properties = collapsed_properties(model_spec, swagger_spec)
```
Comment (analogue) [inline bravado_core/model.py]: I don't see any changes to `model_constructor_test.py` which exercises a `model_spec` with `allOf`

File: bravado_core/model.py [hunk 5]
```diff
@@ -165,17 +175,19 @@ def model_constructor(model, model_spec, constructor_kwargs):
...
+def create_model_repr(model, model_spec, swagger_spec):
     """Generates the repr string for the model.
 
     :param model: Instance of a model
     :param model_spec: model specification
     :type model_spec: dict
+    :param swagger_spec: :class:`bravado_core.spec.Spec`
     :returns: repr string for the model
     """
+    properties = collapsed_properties(model_spec, swagger_spec)
```
Comment (analogue) [inline bravado_core/model.py]: I don't see any changes to `create_model_repr_test.py` which exercises a `model_spec` with `allOf`

File: tests/schema/get_spec_for_prop_test.py [hunk 6]
```diff
@@ -28,6 +28,25 @@ def address_spec():
 
 
 @pytest.fixture
+def bus_address_spec():
+    return {
+        'type': 'object',
```
Comment (analogue) [inline tests/schema/get_spec_for_prop_test.py]: Shouldn't this move to just before line 39?

File: tests/schema/get_spec_for_prop_test.py [hunk 7]
```diff
@@ -28,6 +28,25 @@ def address_spec():
 
 
 @pytest.fixture
+def bus_address_spec():
```
Comment (analogue) [inline tests/schema/get_spec_for_prop_test.py]: Consistency with the `business_address` fixture?  s/bus/business/

File: tests/schema/get_spec_for_prop_test.py [hunk 8]
```diff
@@ -92,6 +121,23 @@ def test_additionalProperties_not_dict_like(minimal_swagger_spec, address_spec,
     assert "Don't know what to do" in str(excinfo.value)
 
 
+def test_polymorphic_object(minimal_swagger_dict, address_spec, address,
```
Comment (analogue) [inline tests/schema/get_spec_for_prop_test.py]: This looks more like a test of the composition via `allOf`.  `test_composition`?

File: tests/unmarshal/unmarshal_object_test.py [hunk 9]
```diff
@@ -117,6 +117,78 @@ def test_with_ref(minimal_swagger_dict, address_spec, location_spec):
     assert result == address
 
 
+def test_with_poly_model(minimal_swagger_dict, address_spec):
```
Comment (analogue) [inline tests/unmarshal/unmarshal_object_test.py]: `test_with_model_composition`?

File: tests/unmarshal/unmarshal_object_test.py [hunk 10]
```diff
@@ -117,6 +117,78 @@ def test_with_ref(minimal_swagger_dict, address_spec, location_spec):
...
+
+    bus_address_spec = {
+        'type': 'object',
```
Comment (analogue) [inline tests/unmarshal/unmarshal_object_test.py]: Move to line 129

File: tests/unmarshal/unmarshal_object_test.py [hunk 11]
```diff
@@ -117,6 +117,78 @@ def test_with_ref(minimal_swagger_dict, address_spec, location_spec):
...
+                }
+            }
+        }
+    }
+    minimal_swagger_dict['paths']['/foo'] = business_address_response
+
+    swagger_spec = Spec.from_dict(minimal_swagger_dict)
+
+    address_dict = {
+        'company': 'n/a',
```
Comment (analogue) [inline tests/unmarshal/unmarshal_object_test.py]: This is subtle, but why are you passing `company` in if it is treated as an `additionalProperty`?  I would expect the test case for an address wouldn't want to cause any confusion with the test case for a business address which expects (but doesn't require) `company`.

File: tests/unmarshal/unmarshal_object_test.py [hunk 12]
```diff
@@ -117,6 +117,78 @@ def test_with_ref(minimal_swagger_dict, address_spec, location_spec):
...
+    }
+    expected_address = {
+        'company': 'n/a',
+        'number': 1600,
+        'street_name': 'Pennsylvania',
+        'street_type': 'Avenue'
+    }
+
+    address = unmarshal_object(swagger_spec, address_spec, address_dict)
+    assert expected_address == address
```
Comment (analogue) [inline tests/unmarshal/unmarshal_object_test.py]: I don't see how what assertion and the related code have to do with unmarshalling a model using composition.

File: tests/unmarshal/unmarshal_model_test.py [hunk 13]
```diff
@@ -5,6 +5,28 @@
 from bravado_core.spec import Spec
 
 
+def test_definitions_with_ref(simple_crossref_spec):
```
Comment (analogue) [inline tests/unmarshal/unmarshal_model_test.py]: This test fixture pollutes `test-data/2.0/simple_crossref/swagger.json` which was previously only used by `from_dict_test.py`. Keeping the test data isolated between tests makes it easier to maintain and reduces weird side effects :-)

**Comment by analogue** [comment]
I didn't see anything that exercises the validation side of things (vs marshalling and unmarshalling which you covered). Some verifications in `validate_object_test.py` would be good.

**Comment by lanzz** [comment]
Looks like this PR only treats properties as composite, e.g. there's no way to combine the `required` lists of the constituent schemas?

File: bravado_core/schema.py [hunk 14]
```diff
@@ -93,3 +96,34 @@ def get_spec_for_prop(swagger_spec, object_spec, object_value, prop_name):
...
+
+    # properties may or may not be present
+    if 'properties' in model_spec:
+        for attr, attr_spec in iteritems(model_spec['properties']):
+            properties[attr] = attr_spec
+
+    # allOf may or may not be present
+    if 'allOf' in model_spec:
+        for item_spec in model_spec['allOf']:
+            deref = swagger_spec.deref
```
Comment (lanzz) [inline bravado_core/schema.py]: This assignment should happen outside of this loop

**Comment by macisamuele** [comment]
Merged in #110, which is this change plus some validation tests.
@captin411  thanks a lot for working on the new feature

**Pull request closed without merging**